openedx/core/djangoapps/content/search/models.py

```diff
@@ -0,0 +1,56 @@
+"""Database models for content search"""
+
+from __future__ import annotations
+
+from typing import List
+
+from django.db import models
+from django.utils.translation import gettext_lazy as _
+from opaque_keys.edx.django.models import LearningContextKeyField
+from rest_framework.request import Request
+
+from cms.djangoapps.contentstore.views.course import get_courses_accessible_to_user
+from openedx.core.djangoapps.content_libraries.api import get_libraries_for_user
+
+
+class SearchAccess(models.Model):
+    """
+    Stores a numeric ID for each ContextKey.
+
+    We use this shorter ID instead of the full ContextKey when determining a user's access to search-indexed course and
+    library content because:
+
+    a) in some deployments, users may be granted access to more than 1_000 individual courses, and
+    b) the search filter request is stored in the JWT, which is limited to 8Kib.
+    """
+    id = models.BigAutoField(
+        primary_key=True,
+        help_text=_(
+            "Numeric ID for each Course / Library context. This ID will generally require fewer bits than the full "
+            "LearningContextKey, allowing more courses and libraries to be represented in content search filters."
+        ),
+    )
+    context_key = LearningContextKeyField(
+        max_length=255, unique=True, null=False,
+    )
+
+
+def get_access_ids_for_request(request: Request) -> List[int]:
+    """
+    Returns the SearchAccess.id values that the user has read access to.
+    """
+    courses, _ = get_courses_accessible_to_user(request)
```

Comment:
I think we want to get the IDs only for course-level assignments here. If the user is staff in `orgA` and doesn't have any course-level role, this list should be empty. I'm not sure if we have a function for that.

Maybe use what we did in `_get_course_user_orgs` in the tagging rules, but return the course ids?

Reply:
Ugh.. I really don't want `content.search` to have complicated access rules like `content_tagging` does.. I was really hoping to re-use existing method instead of having to write our own, and every time I have to write one, I get grumpy that the platform rules aren't already clean and easy to use!

Anyway. I can refactor the content_tagging rules and use what I can from there.

Reply:
Done with [6a3c30a](https://github.com/openedx/edx-platform/pull/34471/commits/6a3c30aa71b83e385e6f5e698187b33c7cbff1e3).